### ISSUE TITLE
Raytracer crash with a blocker

### ISSUE THREAD

**Issue by dingzeyuli** [EVERTims/evertims#3]
Hi, when I try to do the auralization in a scene like this, the raytracer keeps crashing. The only new element in this scene is the center blocker. I tried to increase the reflection order, it seems to help a little bit, but overall the IMS module still crashes when I move the source around the blocker. 

Usually the plugin can be fixed by simply stopping and starting the raytracing module and the on-the-fly aurualization. Do you guys know what might cause this problem? Any possible loophole in the raytracer in the case of multiple bounces? Thank you!

![image](https://cloud.githubusercontent.com/assets/5297902/25635902/a7fe1eb2-2f4d-11e7-94f1-b34f1aa42cb5.png)

[ims_2017-05-02-153952_pixels.crash.txt](https://github.com/EVERTims/evertims/files/971795/ims_2017-05-02-153952_pixels.crash.txt)

[blocker.blend.zip](https://github.com/EVERTims/evertims/files/971800/blocker.blend.zip)

**Comment by PyrApple** [comment]
Indeed, and a hard one (most likely related to multithread based on the logs). searching...

**Comment by dingzeyuli** [comment]
Thanks for the feedback! When I move the source or listener slowly, the raytracer crashes much less frequently. I guess if the previous multi-threaded raytracing has not completed, new movement happens and triggers a new round of rays. This might have led to the crashing. 

Besides the geometry built-in to blender, can I import custom shapes? I made a mesh with OpenSCAD and it seems the raytracer is not behaving correctly. Is there any requirement for the input mesh? mesh quality? topology limit?

[office.stl.zip](https://github.com/EVERTims/evertims/files/973833/office.stl.zip)

![image](https://cloud.githubusercontent.com/assets/5297902/25667978/3be686a0-2ff3-11e7-891b-5f9b3c3a55cb.png)

**Comment by dingzeyuli** [comment]
@PyrApple: 
After some trials, I realize the raytracer handles quad meshes much better than tri meshes. At least, no more penetration happens after I switch to quad mesh. 

Yet, in the case of the a door and a hallway, the engine still fails. It either computes a wrong acoustic decay or crashes again like in the triangle mesh cases. I was wondering if you have some already-verified complex scene that I can play with? Something more complex than the square room? Thank you very much.

**Comment by PyrApple** [comment]
Quad vs Tri mesh, you're definitely right there. No official complex model that I can share for the moment, yet would be pleased to handle one of yours that highlights a bug in acoustic decay estimation. Would you mind opening a new thread on that and keep this one for the blocker issue?

**Comment by dingzeyuli** [comment]
@PyrApple  Okay! I will upload my slightly more complex model in a new thread.

Since you have worked on this acoustic simulator for a while, are you aware of other similar project or software that has similar functionalities? Thanks a lot!

**Comment by PyrApple** [comment]
Non-exhaustive listing of auralization / spatialization software: [EVERTims page on my personal website](https://pyrapple.github.io/pages/evertims.html).
